### ISSUE TITLE
Grep argument list too long when running overcommit -r with MergeConflicts

### ISSUE THREAD

**Issue by donaldpiret** [brigade/overcommit#226]
Getting the following error when trying to run overcommit -r 

```
Checking for merge conflicts.........................[MergeConflicts] FAILED
Hook raised unexpected error
Argument list too long - grep
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/childprocess-0.5.6/lib/childprocess/unix/fork_exec_process.rb:58:in `launch_process'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/childprocess-0.5.6/lib/childprocess/abstract_process.rb:82:in `start'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/subprocess.rb:23:in `spawn'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/utils.rb:142:in `execute'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/hook/base.rb:111:in `execute'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/hook/pre_commit/merge_conflicts.rb:5:in `run'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/hook/base.rb:45:in `run_and_transform'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/hook_runner.rb:107:in `block in run_hook'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/interrupt_handler.rb:63:in `disable_until_finished_or_interrupted'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/hook_runner.rb:106:in `run_hook'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/hook_runner.rb:59:in `block in run_hooks'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/hook_runner.rb:58:in `each'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/hook_runner.rb:58:in `run_hooks'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/hook_runner.rb:39:in `block in run'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/interrupt_handler.rb:84:in `isolate_from_interrupts'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/hook_runner.rb:24:in `run'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/cli.rb:193:in `run_all'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/lib/overcommit/cli.rb:25:in `run'
/Users/donald/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.25.0/bin/overcommit:8:in `<top (required)>'
/Users/donald/.rbenv/versions/2.1.5/bin/overcommit:23:in `load'
/Users/donald/.rbenv/versions/2.1.5/bin/overcommit:23:in `<main>'
```

**Comment by sds** [comment]
Hey @donaldpiret, thanks for the report.

You're likely trying to commit a large number of files. Depending on the OS you're running (looks like Mac OS X) you're limited to a command line of 262144 characters&mdash;run `getconf ARG_MAX` on your machine to confirm.

I'm not sure this is a common-enough use case to warrant changing the implementing to search through each file serially, though I'll acknowledge it would probably not be that much slower. The main reason we use `grep` is it provides an easy way to exclude binary files and is faster than opening and searching the file in Ruby.

I'll have a look at some point to see if there's a significant performance penalty between using `grep` and using Ruby. If in the typical case of committing a few files it's negligible, then I'd consider switching away from `grep` to Ruby (though it would likely be much slower for large file lists like in your situation).

**Comment by donaldpiret** [comment]
Perhaps when running with the -r flag the whole project directory could be passed instead of each individual file? Would this be a way of avoiding the issue?

**Comment by jawshooah** [comment]
That would be difficult to implement in a general way, since a lot of the hooks operate on specific file types and don't know how to process other kinds of files.

**Comment by sds** [comment]
I think Overcommit should provide an abstraction around `xargs` which allows you to chunk large argument lists together for tools where it makes sense to.

Quick proof of concept (I tested this on a directory with ~28K files):

```ruby
require 'tempfile'

Tempfile.new('args').tap do |f|
  files = Dir['**/*'].reject { |file| File.directory?(file) }

  files.each do |file|
    f.puts(file)
  end
  f.close

  puts `xargs --arg-file=#{f.path} grep -IHn "Some search string"`
end
```

This took ~400ms, so performance is respectable.

Key point is that you can pass an argument file to `xargs` which will then chunk them based on your system's `ARG_MAX` and invoke specified command as many times as necessary, grouping the output back into a single stream. We can provide a wrapper around this so hook authors don't need to worry about the details.

I imagine we'll probably want to modify the `execute` helper to take a second argument that allows you to specify the list of file arguments separately.

I'll get to it when I can, but if there are questions concerns about this approach let me know.

**Comment by seanknox** [comment]
@sds The MergeConflict hook fails with no files in my changeset.

```
$ git status                                                                                                                                                       
On branch master
nothing to commit, working directory clean

$ overcommit -r
Running pre-commit hooks
Checking for merge conflicts.........................[MergeConflicts] FAILED
Hook raised unexpected error
Argument list too long - grep
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/childprocess-0.5.6/lib/childprocess/unix/fork_exec_process.rb:58:in `launch_process'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/childprocess-0.5.6/lib/childprocess/abstract_process.rb:82:in `start'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/subprocess.rb:23:in `spawn'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/utils.rb:142:in `execute'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/hook/base.rb:86:in `execute'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/hook/pre_commit/merge_conflicts.rb:5:in `run'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/hook/base.rb:45:in `block in run_and_transform'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/utils.rb:170:in `with_environment'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/hook/base.rb:45:in `run_and_transform'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/hook_runner.rb:107:in `block in run_hook'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/interrupt_handler.rb:63:in `disable_until_finished_or_interrupted'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/hook_runner.rb:106:in `run_hook'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/hook_runner.rb:59:in `block in run_hooks'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/hook_runner.rb:58:in `each'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/hook_runner.rb:58:in `run_hooks'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/hook_runner.rb:39:in `block in run'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/interrupt_handler.rb:84:in `isolate_from_interrupts'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/hook_runner.rb:24:in `run'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/cli.rb:193:in `run_all'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/lib/overcommit/cli.rb:25:in `run'
/opt/rubies/2.1.5/lib/ruby/gems/2.1.0/gems/overcommit-0.26.0/bin/overcommit:8:in `<top (required)>'
/opt/boxen/rbenv/versions/2.1.5/bin/overcommit:23:in `load'
/opt/boxen/rbenv/versions/2.1.5/bin/overcommit:23:in `<main>'

✗ One or more pre-commit hooks failed
```

With `MergeConflict` disabled, no error.

**Comment by jawshooah** [comment]
From the README:

> `-r/--run` | Run pre-commit hook against all tracked files in repository

The `--run` flag treats all files in the repository as staged, which is why the `MergeConflicts` hook currently fails with `argument list too long` for very large repositories. The changes @sds proposed above would solve that problem.

**Comment by sds** [comment]
Fixed in 82d5220. 

I'm not sure how robust the solution I've come up with is, but it at least doesn't break existing behavior. Since this is likely only a problem for hooks that check _all_ files rather than specific ones, I won't be modifying other hooks to use the pattern just yet since they're unlikely to benefit from it, and it doesn't come with a downside of making dealing with exit statuses more difficult (since we're dealing with multiple exit statuses rather than just one).
